File: java/3267. Count Almost Equal Pairs II.java
class Solution {
    public int countPairs(int[] nums) {
        Arrays.sort(nums);
        Map<Integer, Integer> freq = new HashMap<>();
        int answer = 0;
        for (int i = 0; i < nums.length; i++) {
            Set<Integer> swaps = getAll(nums[i]);
            Set<Integer> total = new HashSet<>();
            for (int j : swaps) {
                for (int k : getAll(j)) {
                    total.add(k);
                }
            }
            for (int j : total) {
                answer += freq.getOrDefault(j, 0);
            }
            freq.put(nums[i], freq.getOrDefault(nums[i], 0) + 1);
        }
        return answer;
    }
    private Set<Integer> getAll(int num) {
        char[] arr = String.valueOf(num).toCharArray();
        Set<Integer> possible = new HashSet<>();
        possible.add(num);
        for (int i = 0; i < arr.length; i++) {
            for (int j = i+1; j < arr.length; j++) {
                char temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
                possible.add(Integer.parseInt(String.valueOf(arr)));
                temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
        return possible;
    }
}

File: go/3267. Count Almost Equal Pairs II.go
func countPairs(nums []int) int {
    sort.Ints(nums)
    freq := make(map[int]int)
    answer := 0
    for _, num := range nums {
        swap := getSwaps(num)
        total := make(map[int]bool)
        for j := range swap {
            for k := range getSwaps(j) {
                total[k] = true
            }
        }
        for i := range total {
            answer += freq[i]
        }
        freq[num]++
    }
    return answer
}

func getSwaps(num int) map[int]bool {
    result := make(map[int]bool)
    result[num] = true
    digits := getDigits(num)
    for i := 0; i < len(digits); i++ {
        for j := i+1; j < len(digits); j++ {
            digits[i], digits[j] = digits[j], digits[i]
            result[getVal(digits)] = true
            digits[i], digits[j] = digits[j], digits[i]
        }
    }
    return result
}

func getDigits(num int) []int {
    answer := []int{}
    for num > 0 {
        answer = append(answer, num % 10)
        num /= 10
    }
    return answer
}

func getVal(digits []int) int {
    answer := 0;
    for i := len(digits)-1; i >= 0; i-- {
        answer *= 10
        answer += digits[i]
    }
    return answer
}

